accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -263,6 +268,157 @@ The **flatten** function is modified as follows:
+to "return void to null" in an `async` function, nor to "return a void future
+to null"._
+
+The next sentence is changed as follows:
+
+```
+It is a compile-time error if $s$ is \code{\RETURN{} $e$;},
+\flatten{S} is not \VOID,
+$S$ is not assignable to $T$,
+and flatten{S} is not a subtype of $T$.
+```
+
+_Comparing to Dart before null-safety, this means that it is now allowed
+to return a future when the future value type is a suitable future; for
+instance, we can have `return Future<int>.value(42)` in an `async` function
+with declared return type `Future<Future<int>>`. Conversely, it is no longer
+allowed to return a `Future<dynamic>` or `FutureOr<dynamic>` when the future
+value type is `Future<U>` for some `U` which is not a top type, because it is
+no longer enough that `dynamic` is assignable to `U`._
+
+The dynamic semantics specified at
+[this location](https://github.com/dart-lang/language/blob/65b8267be0ebb9b3f0849e2061e6132021a4827d/specification/dartLangSpec.tex#L15597)
+is changed as follows, where `$f$` is the enclosing function with declared
+return type `$T$`, and `$e$` is the returned expression with static type `$S$`:
+
+```
+When $f$ is a synchronous non-generator, evaluation proceeds as follows:
+The expression $e$ is evaluated to an object $o$.
+A dynamic error occurs unless the dynamic type of $o$ is a subtype of
+the actual return type of $f$
+(\ref{actualTypes}).
+Then the return statement $s$ completes returning $o$
+(\ref{statementCompletion}).
+
+\commentary{%
+The case where the evaluation of $e$ throws is covered by
+the general rule which propagates the throwing completion,
+in this case to the call site.%
+}
+
+When $f$ is an asynchronous non-generator with future value type $T_v$
+(\ref{functions}),
+the return statement $s$ is desugared at compile time as described below.
+The resulting code uses \code{\_return $e'$},
+which is a statement that evaluates $e'$ to an object
+$o'$, completes the returned future of $f$ with $o'$,
+and finally returns without an object
+(\ref{statementCompletion}),
+thus terminating the execution of $f$.
+Alternatively, if the evaluation of $e'$ throws
+an object \metavar{error} and stack trace \metavar{stackTrace},
+the returned future of $f$ is error-completed
+with \metavar{error} and \metavar{stackTrace},
+and finally the statement returns without an object.
+The desugaring proceeds as follows,
+where $\Gamma$ is the typing environment of $s$,
+and the first item whose requirement is satisfied is used:
+
+\begin{itemize}
+\item% 1
+  When $S$ is \DYNAMIC:
+  \code{\VAR r = $e$; \_return r \IS{} Future<$T_v$> ? \AWAIT{} r : (r \AS{} $T_v$);}
+\item% 2
+  When \SubtypeStd{S}{T_v}: \code{\_return $e$;}
+\item% 3
+  When \SubtypeStd{S}{\code{Future<$T_v$>}}: \code{\_return \AWAIT{} $e$;}
+\item% 4
+  When $S$ is of the form \code{FutureOr<$U$>} where \SubtypeStd{U}{T_v}:
+  \code{\VAR r = $e$; \_return r is Future<$T_v$> ? \AWAIT{} r : r;}
```

Comment:
The reason why I changed it to `Future<T_v>` is that this is obviously not changing the semantics (compared to today).

So it would be good to know which cases would make a difference, and whether the beauty is worth the breakage. Do you have an example where the beauty is evident? ;-)

Reply:
Newest rules do not have this case.